hpx/util/checkpoint.hpp

```diff
@@ -0,0 +1,516 @@
+// Copyright (c) 2017 Adrian Serio
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+/// This header defines the save_checkpoint and restore_checkpoint functions.
+/// These functions are designed to help HPX application developer's checkpoint
+/// their applications. Save_checkpoint serializes one or more objects and saves
+/// them as a byte stream. Restore_checkpoint converts the byte stream back into
+/// instances of the objects.
+//
+
+/// \file hpx/util/checkpoint.hpp
+
+#if !defined(CHECKPOINT_HPP_07262017)
+#define CHECKPOINT_HPP_07262017
+
+#include <hpx/dataflow.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/runtime/serialization/serialize.hpp>
+#include <hpx/runtime/serialization/vector.hpp>
+
+#include <cstddef>
+#include <fstream>
+#include <iosfwd>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace hpx {
+namespace util {
+
+    // Forward declarations
+    class checkpoint;
+    std::ostream& operator<<(std::ostream& ost, checkpoint const& ckp);
+    std::istream& operator>>(std::istream& ist, checkpoint& ckp);
+    namespace detail {
+        struct save_funct_obj;
+    }
+
+    ///////////////////////////////////
+    /// Checkpoint Object
+    ///
+    /// Checkpoint is the container object which is produced by save_checkpoint
+    /// and is consumed by a restore_checkpoint. A checkpoint may be moved into
+    /// the save_checkpoint object to write the byte stream to the pre-created
+    /// checkpoint object.
+    class checkpoint
+    {
+        std::vector<char> data;
+
+        friend std::ostream& operator<<(
+            std::ostream& ost, checkpoint const& ckp);
+        friend std::istream& operator>>(std::istream& ist, checkpoint& ckp);
+        //Serialization Definition
+        friend class hpx::serialization::access;
+        template <typename Archive>
+        void serialize(Archive& arch, const unsigned int version)
+        {
+            arch& data;
+        };
+        friend struct detail::save_funct_obj;
+        template <typename T, typename... Ts>
+        friend void restore_checkpoint(checkpoint const& c, T& t, Ts&... ts);
+
+    public:
+        checkpoint() = default;
+        checkpoint(checkpoint const& c)
+          : data(c.data)
+        {
+        }
+        checkpoint(checkpoint&& c)
+          : data(std::move(c.data))
+        {
+        }
+        ~checkpoint() = default;
+
+        //Other Constructors
+        checkpoint(char* stream, std::size_t count)
+        {
+            for (std::size_t i = 0; i < count; i++)
+            {
+                data.push_back(*stream);
+                stream++;
+            }
```

Comment:
You should resize the vector and then memcpy the data into it here

Reply:
Yes, I agree - the signatures of these should be:

    checkpoint(std::vector<char> const& stream);
    checkpoint(std::vector<char> && stream);

or if you really must,

    template <typename InIter>
    checkpoint(InIter begin, InIter end);

Reply:
@hkaiser and I are discussing options to remove/replace this construction entirely as exposing pointers in the interface is a weak design.